File: GaiaXAndroid/src/main/kotlin/app/visly/stretch/Layout.kt
package app.visly.stretch

/**
 * @suppress
 */
data class Layout(
    var x: Float,
    var y: Float,
    val width: Float,
    val height: Float,
    val children: MutableList<Layout>,
    var id: String = ""
) {

    companion object {
        fun fromFloatArray(args: FloatArray, offset: Int): Pair<Int, Layout> {
            var offset = offset

            val x = args[offset++]
            val y = args[offset++]
            val width = args[offset++]
            val height = args[offset++]
            val childCount = args[offset++].toInt()
            val children = mutableListOf<Layout>()

            for (i in 0 until childCount) {
                val child = Layout.fromFloatArray(args, offset)
                offset = child.first
                children.add(child.second)
            }

            return Pair(offset, Layout(x, y, width, height, children))
        }
    }

    override fun toString(): String {
        return "Layout(x=$x, y=$y, width=$width, height=$height, children=$children, id='$id')"
    }

}

File: GaiaXHarmony/GXStretchBinding/GXStretch/src/main/ets/stretch/Layout.ts
import { ArrayList } from '@kit.ArkTS';
import { hilog } from '@kit.PerformanceAnalysisKit';

class Tmp {
  offset: number
  layout: Layout

  constructor(offset, layout) {
    this.offset = offset;
    this.layout = layout;

  }
}

export class Layout {
  x: number
  y: number
  width: number
  height: number
  children: ArrayList<Layout>

  constructor(x: number, y: number, width: number, height: number, children: ArrayList<Layout>) {
    this.x = x
    this.y = y
    this.width = width
    this.height = height
    this.children = children
  }

  static fromFloatArray(args: number[], off: number) {
    let offset = off;
    let x = args[offset++];
    let y = args[offset++];
    let width = args[offset++];
    let height = args[offset++];
    let childCount = args[offset++];

    hilog.info(0x0000, 'GXStretch', `Layout fromFloatArray offset=${offset} x=${x} y=${y} width=${width} height=${height} childCount=${childCount}`);

    let children = new ArrayList<Layout>()
    for (let i = 0; i < childCount; i++) {
      let child = Layout.fromFloatArray(args, offset)
      offset = child.offset;
      children.add(child.layout);
    }
    return new Tmp(offset, new Layout(x, y, width, height, children));
  }

  toString() {
    return `Layout(width=${this.width} height=${this.height} x=${this.x} y=${this.y} children.length=${this.children.length})`
  }
}